### PR TITLE
[GSoC] Added sampling for Joint Random Variables

### LINKED ISSUE #17622: Sampling from joint normal distribution fails
The following code that would sample from a joint normal distribution fails:

```python
import sympy.stats
X = sympy.stats.Normal('X', [0, 0], [[0.1, 0.025], [0.025, 0.1]])
sympy.stats.sample(X)
```

The latter part of the error message:
```
  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 78, in domain
    return SingleDomain(self.symbol, self.set)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 50, in set
    return self.domain.set

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/joint_rv.py", line 76, in domain
    rvs = random_symbols(self.distribution)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/stats/rv.py", line 510, in random_symbols
    return list(atoms(RandomSymbol))

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/core/basic.py", line 507, in atoms
    for expr in preorder_traversal(self):

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/core/basic.py", line 2007, in __next__
    return next(self._pt)

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/core/basic.py", line 1978, in _preorder_traversal
    for subtree in self._preorder_traversal(arg, keys):

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/core/basic.py", line 1978, in _preorder_traversal
    for subtree in self._preorder_traversal(arg, keys):

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/core/basic.py", line 1978, in _preorder_traversal
    for subtree in self._preorder_traversal(arg, keys):

  File "/home/rikard/.local/lib/python3.7/site-packages/sympy/core/basic.py", line 1965, in _preorder_traversal
    if isinstance(node, Basic):

RecursionError: maximum recursion depth exceeded while calling a Python object
```

I am using sympy 1.4 and python 3.7.3

### PR DESCRIPTION
**Pull request by Smit-create**
`sample` is made to support sampling from Joint RVs
#### References to other Issues or PRs

Fixes #17622

#### Release Notes

* stats
  * Added sampling of Joint RVs from external libraries

ping @czgdp1807 @Upabjojr @jmig5776

### CODE REVIEW

**Comment by czgdp1807** [comment]
Can you show a working example?

**Comment by Smit-create** [comment]
> Can you show a working example?

From issue #17622
```py
>>> import sympy.stats
>>> X = sympy.stats.Normal('X', [0, 0], [[0.1, 0.025], [0.025, 0.1]])
>>> itr = sympy.stats.sample(X)
>>> next(itr)
array([-0.31544473,  0.31495318])
>>> itr = sympy.stats.sample(X, size=3)
>>> next(itr)
array([[ 0.05231316, -0.38950943],
       [ 0.25917453,  0.43911207],
       [ 0.21743206,  0.37448919]])
```

**Comment by Upabjojr** [comment]
what about adding support to tensorflow.probability? I've already commented on the other issue.

For example: https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/JointDistributionSequential

**Comment by Upabjojr** [comment]
> Can you show a working example?

Maybe a working example should be added to the documentation.

**Comment by Smit-create** [comment]
> what about adding support to tensorflow.probability

We can surely support that too. The current sample has three library options of `scipy`, `numpy` and `pymc3`. We can also add the 4th option of `tensorflow`. But, in doing so we should make sure that it works for all FRV, DRV, and CRV as well along with JRV. Adding a new library option can be done for all the RVs in new PR after completing the three libraries for JRV here and then adding `tensorflow` for all the four RV i.e. FRV, DRV, CRV and JRV.

**Comment by Smit-create** [comment]
I have added sampling of Multivariate Normal, Multivariate Beta(Dirichlet), and Multinomial Distributions from `scipy`, `numpy`, and `pymc3`. One can easily extend with more distributions(currently I found these 3 in their docs) by just adding the Distribution to their respective library's dictionary and its conversion to external library from `sympy`

File: sympy/stats/rv.py [hunk 1]
```diff
@@ -1040,6 +1040,15 @@ def sample(expr, condition=None, size=(), library='scipy', numsamples=1,
...
     >>> [sam in N.pspace.domain.set for sam in samp_list] # doctest: +SKIP
     [True, True, True, True]
+    >>> G = Geometric('G', 0.5) # Discrete Random Variable
+    >>> samp_list = next(sample(G, size=3)) # doctest: +SKIP
+    >>> [sam in G.pspace.domain.set for sam in samp_list] # doctest: +SKIP
+    [True, True, True]
+    >>> MN = Normal("MN", [3, 4], [[2, 1], [1, 2]]) # Joint Random Variable
+    >>> samp_list = next(sample(MN, size=4)) # doctest: +SKIP
+    >>> [tuple(sam) in MN.pspace.domain.set for sam in samp_list] # doctest: +SKIP
+    [True, True, True, True]
```
Comment (czgdp1807) [inline sympy/stats/rv.py:1050]: Numbers would be better i.e., samp_list.

**Comment by Upabjojr** [comment]
> > what about adding support to tensorflow.probability
> 
> We can surely support that too. The current sample has three library options of `scipy`, `numpy` and `pymc3`. We can also add the 4th option of `tensorflow`. But, in doing so we should make sure that it works for all FRV, DRV, and CRV as well along with JRV. Adding a new library option can be done for all the RVs in new PR after completing the three libraries for JRV here and then adding `tensorflow` for all the four RV i.e. FRV, DRV, CRV and JRV.

Yes, definitely in a new PR. Anyway, you don't need to implement all methods, you can just create a method calling `raise NotImplementedError("backend to tensorflow probability not yet available for this distribution")` in the common superclass, then implement the methods in the subclasses when the available time allows it.

**Comment by czgdp1807** [comment]
Adding new libraries is an easy task now. Probably, let's leave the task of adding tensorflow support for some other time. Open an issue for it and we can look at it later.

**Comment by czgdp1807** [comment]
Will be merged tomorrow if no objections raised. @Upabjojr @jmig5776

**Pull request merged**